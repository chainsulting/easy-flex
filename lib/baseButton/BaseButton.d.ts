import React, { ButtonHTMLAttributes } from 'react';
import { BorderProps } from '../utils/border';
import { ColorProps } from '../utils/color';
import { DistanceProps } from '../utils/distance';
import { FlexContainerProps } from '../utils/flexContainer';
import { FlexItemProps } from '../utils/flexItem';
import { FontProps } from '../utils/font';
import { OverflowProps } from '../utils/overflow';
import { SizeProps } from '../utils/size';
export declare type BaseButtonProps = Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'color'> & BorderProps & ColorProps & DistanceProps & FlexContainerProps & FlexItemProps & FontProps & OverflowProps & SizeProps;
export declare type ExternalBaseButtonProps = Omit<BaseButtonProps, 'align' | 'backgroundColor' | 'borderColor' | 'borderRadius' | 'borderStyle' | 'borderWidth' | 'color' | 'direction' | 'fontFamily' | 'fontSize' | 'fontWeight' | 'gap' | 'italic' | 'justify' | 'lineHeight' | 'overflow' | 'overflowX' | 'overflowY' | 'padding' | 'paddingBottom' | 'paddingHorizontal' | 'paddingLeft' | 'paddingRight' | 'paddingTop' | 'paddingVertical' | 'round' | 'underline' | 'underlineColor'>;
export declare const BaseButton: React.ForwardRefExoticComponent<Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, "color"> & BorderProps & ColorProps & import("..").MarginProps & import("..").PaddingProps & FlexContainerProps & FlexItemProps & FontProps & OverflowProps & SizeProps & React.RefAttributes<HTMLButtonElement>>;
