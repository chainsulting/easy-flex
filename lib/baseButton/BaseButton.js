"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseButton = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var react_1 = require("react");
var styled_components_1 = __importDefault(require("styled-components"));
var border_1 = require("../utils/border");
var color_1 = require("../utils/color");
var distance_1 = require("../utils/distance");
var flexContainer_1 = require("../utils/flexContainer");
var flexItem_1 = require("../utils/flexItem");
var font_1 = require("../utils/font");
var overflow_1 = require("../utils/overflow");
var size_1 = require("../utils/size");
var StyledButton = styled_components_1.default.button(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n\tbox-sizing: border-box;\n\tdisplay: flex;\n\tborder: none;\n\tcursor: pointer;\n\t", "\n\t", "\n\t", "\n\t", "\n\t", "\n\t", "\n\t", "\n\t", "\n\t\n\t&:disabled {\n\t\tcursor: not-allowed;\n\t}\n\n\t&:focus:not(:focus-visible) {\n\t\toutline: none;\n\t}\n"], ["\n\tbox-sizing: border-box;\n\tdisplay: flex;\n\tborder: none;\n\tcursor: pointer;\n\t", "\n\t", "\n\t", "\n\t", "\n\t", "\n\t", "\n\t", "\n\t", "\n\t\n\t&:disabled {\n\t\tcursor: not-allowed;\n\t}\n\n\t&:focus:not(:focus-visible) {\n\t\toutline: none;\n\t}\n"])), border_1.borderStyle, color_1.colorStyle, distance_1.distanceStyle, flexContainer_1.flexContainerStyle, flexItem_1.flexItemStyle, font_1.fontStyle, overflow_1.overflowStyle, size_1.sizeStyle);
exports.BaseButton = (0, react_1.forwardRef)(function (_a, ref) {
    var align = _a.align, alignSelf = _a.alignSelf, _b = _a.backgroundColor, backgroundColor = _b === void 0 ? 'transparent' : _b, basis = _a.basis, borderColor = _a.borderColor, borderRadius = _a.borderRadius, borderStyle = _a.borderStyle, borderWidth = _a.borderWidth, children = _a.children, color = _a.color, flex = _a.flex, _c = _a.direction, direction = _c === void 0 ? 'row' : _c, fontFamily = _a.fontFamily, fontSize = _a.fontSize, fontWeight = _a.fontWeight, fullHeight = _a.fullHeight, fullWidth = _a.fullWidth, gap = _a.gap, grow = _a.grow, height = _a.height, italic = _a.italic, justify = _a.justify, lineHeight = _a.lineHeight, margin = _a.margin, marginBottom = _a.marginBottom, marginHorizontal = _a.marginHorizontal, marginLeft = _a.marginLeft, marginRight = _a.marginRight, marginTop = _a.marginTop, marginVertical = _a.marginVertical, maxHeight = _a.maxHeight, maxWidth = _a.maxWidth, minHeight = _a.minHeight, minWidth = _a.minWidth, overflow = _a.overflow, overflowX = _a.overflowX, overflowY = _a.overflowY, padding = _a.padding, paddingBottom = _a.paddingBottom, paddingHorizontal = _a.paddingHorizontal, paddingLeft = _a.paddingLeft, paddingRight = _a.paddingRight, paddingTop = _a.paddingTop, paddingVertical = _a.paddingVertical, round = _a.round, shrink = _a.shrink, underline = _a.underline, width = _a.width, props = __rest(_a, ["align", "alignSelf", "backgroundColor", "basis", "borderColor", "borderRadius", "borderStyle", "borderWidth", "children", "color", "flex", "direction", "fontFamily", "fontSize", "fontWeight", "fullHeight", "fullWidth", "gap", "grow", "height", "italic", "justify", "lineHeight", "margin", "marginBottom", "marginHorizontal", "marginLeft", "marginRight", "marginTop", "marginVertical", "maxHeight", "maxWidth", "minHeight", "minWidth", "overflow", "overflowX", "overflowY", "padding", "paddingBottom", "paddingHorizontal", "paddingLeft", "paddingRight", "paddingTop", "paddingVertical", "round", "shrink", "underline", "width"]);
    var borderStyleProps = (0, border_1.useBorderStyleProps)({ borderColor: borderColor, borderRadius: borderRadius, borderStyle: borderStyle, borderWidth: borderWidth, round: round });
    var colorStyleProps = (0, color_1.useColorStyleProps)({ backgroundColor: backgroundColor, color: color });
    var distanceStyleProps = (0, distance_1.useDistanceStyleProps)({
        margin: margin,
        marginBottom: marginBottom,
        marginHorizontal: marginHorizontal,
        marginLeft: marginLeft,
        marginRight: marginRight,
        marginTop: marginTop,
        marginVertical: marginVertical,
        padding: padding,
        paddingBottom: paddingBottom,
        paddingHorizontal: paddingHorizontal,
        paddingLeft: paddingLeft,
        paddingRight: paddingRight,
        paddingTop: paddingTop,
        paddingVertical: paddingVertical,
    });
    var fontStyleProps = (0, font_1.useFontStyleProps)({ fontFamily: fontFamily, fontSize: fontSize, fontWeight: fontWeight, italic: italic, lineHeight: lineHeight, underline: underline });
    var flexContainerStyleProps = (0, flexContainer_1.useFlexContainerStyleProps)({ align: align, direction: direction, gap: gap, justify: justify });
    var flexItemStyleProps = (0, flexItem_1.useFlexItemStyleProps)({ alignSelf: alignSelf, basis: basis, flex: flex, grow: grow, shrink: shrink });
    var overflowStyleProps = (0, overflow_1.useOverflowStyleProps)({ overflow: overflow, overflowX: overflowX, overflowY: overflowY });
    var sizeStyleProps = (0, size_1.useSizeStyleProps)({
        fullHeight: fullHeight,
        fullWidth: fullWidth,
        height: height,
        maxHeight: maxHeight,
        maxWidth: maxWidth,
        minHeight: minHeight,
        minWidth: minWidth,
        width: width,
    });
    return ((0, jsx_runtime_1.jsx)(StyledButton, __assign({}, borderStyleProps, colorStyleProps, distanceStyleProps, flexContainerStyleProps, flexItemStyleProps, fontStyleProps, overflowStyleProps, sizeStyleProps, { ref: ref }, props, { children: children })));
});
exports.BaseButton.displayName = 'BaseButton';
var templateObject_1;
