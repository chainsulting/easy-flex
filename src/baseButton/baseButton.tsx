import React, { ButtonHTMLAttributes, FC, useMemo } from 'react';
import styled from 'styled-components';
import {
	IAlignItems,
	IAlignSelf,
	IBorderRadius,
	IBorderWidth,
	IColor,
	IDistance,
	IHeight,
	IJustifyContent,
	IWidth,
} from '../types';
import {
	getBorderRadius,
	getBorderWidth,
	ifNotUndefined,
	toPx,
	useColor,
	useDistance,
	useEasyFlexTheme,
	useSize,
} from '../utils';

const StyledBaseButton = styled.button<{
	'data-align'?: IAlignItems;
	'data-align-self'?: IAlignSelf;
	'data-background-color': string;
	'data-border-color'?: string;
	'data-border-radius'?: string;
	'data-border-style'?: 'solid';
	'data-border-width'?: string;
	'data-color'?: string;
	'data-grow'?: number;
	'data-height'?: string;
	'data-height-max'?: string;
	'data-height-min'?: string;
	'data-justify'?: IJustifyContent;
	'data-margin-bottom': string;
	'data-margin-left': string;
	'data-margin-right': string;
	'data-margin-top': string;
	'data-padding-bottom': string;
	'data-padding-left': string;
	'data-padding-right': string;
	'data-padding-top': string;
	'data-shrink'?: number;
	'data-width'?: string;
	'data-width-max'?: string;
	'data-width-min'?: string;
}>`
	display: flex;
	box-sizing: border-box;
	flex-direction: row;
	border: none;
	cursor: pointer;
	outline: none;
	align-items: ${({ 'data-align': align }) => align};
	align-self: ${({ 'data-align-self': alignSelf }) => alignSelf};
	background-color: ${({ 'data-background-color': backgroundColor }) => backgroundColor};
	border-color: ${({ 'data-border-color': borderColor }) => borderColor};
	border-radius: ${({ 'data-border-radius': borderRadius }) => borderRadius};
	border-style: ${({ 'data-border-style': borderStyle }) => borderStyle};
	border-width: ${({ 'data-border-width': borderWidth }) => borderWidth};
	color: ${({ 'data-color': color }) => color};
	flex-grow: ${({ 'data-grow': grow }) => grow};
	height: ${({ 'data-height': height }) => height};
	max-height: ${({ 'data-height-max': heightMax }) => heightMax};
	min-height: ${({ 'data-height-min': heightMin }) => heightMin};
	justify-content: ${({ 'data-justify': justify }) => justify};
	margin-bottom: ${({ 'data-margin-bottom': marginBottom }) => marginBottom};
	margin-left: ${({ 'data-margin-left': marginLeft }) => marginLeft};
	margin-right: ${({ 'data-margin-right': marginRight }) => marginRight};
	margin-top: ${({ 'data-margin-top': marginTop }) => marginTop};
	padding-bottom: ${({ 'data-padding-bottom': paddingBottom }) => paddingBottom};
	padding-left: ${({ 'data-padding-left': paddingLeft }) => paddingLeft};
	padding-right: ${({ 'data-padding-right': paddingRight }) => paddingRight};
	padding-top: ${({ 'data-padding-top': paddingTop }) => paddingTop};
	flex-shrink: ${({ 'data-shrink': shrink }) => shrink};
	width: ${({ 'data-width': width }) => width};
	max-width: ${({ 'data-width-max': widthMax }) => widthMax};
	min-width: ${({ 'data-width-min': widthMin }) => widthMin};
`;

export interface IBaseButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
	/** The alignment of the component's children on the cross axis. */
	align?: IAlignItems;
	/** The alignment of the component on the parent's element cross axis. */
	alignSelf?: IAlignSelf;
	/** Component's background color. */
	backgroundColor?: IColor;
	/** Component's border color. */
	borderColor?: IColor;
	/** Component's border radius. */
	borderRadius?: IBorderRadius | number;
	/** Component's border width. */
	borderWidth?: IBorderWidth | number;
	/** Component's color. */
	color?: IColor;
	/** Sets the component's height to 100% if true. */
	fullHeight?: boolean;
	/** Sets the component's width to 100% if true. */
	fullWidth?: boolean;
	/** Component's flex grow. */
	grow?: number;
	/** Component's height. */
	height?: IHeight | number;
	/** Sets how the browser distributes space between and around the component's children along the main axis. */
	justify?: IJustifyContent;
	/** Component's margin of all sides. */
	margin?: IDistance | number;
	/** Component's bottom margin. */
	marginBottom?: IDistance | number;
	/** Component's left margin. */
	marginLeft?: IDistance | number;
	/** Component's right margin. */
	marginRight?: IDistance | number;
	/** Component's top margin. */
	marginTop?: IDistance | number;
	/** Component's left and right margin. */
	marginX?: IDistance | number;
	/** Component's top and bottom margin. */
	marginY?: IDistance | number;
	/** Component's maximum height. */
	maxHeight?: IHeight | number;
	/** Component's maximum width. */
	maxWidth?: IWidth | number;
	/** Component's miniumum height. */
	minHeight?: IHeight | number;
	/** Component's minimum width. */
	minWidth?: IWidth | number;
	/** Component's padding of all sides. */
	padding?: IDistance | number;
	/** Component's bottom padding. */
	paddingBottom?: IDistance | number;
	/** Component's left padding. */
	paddingLeft?: IDistance | number;
	/** Component's right padding. */
	paddingRight?: IDistance | number;
	/** Component's top padding. */
	paddingTop?: IDistance | number;
	/** Component's left and right padding. */
	paddingX?: IDistance | number;
	/** Component's top and bottom padding. */
	paddingY?: IDistance | number;
	/** Component's flex shrink. */
	shrink?: number;
	/** Component's width. */
	width?: IWidth | number;
}

export type IExternalBaseButtonProps = Omit<
	IBaseButtonProps,
	| 'align'
	| 'backgroundColor'
	| 'borderColor'
	| 'borderRadius'
	| 'borderWidth'
	| 'color'
	| 'justify'
	| 'padding'
	| 'paddingBottom'
	| 'paddingLeft'
	| 'paddingRight'
	| 'paddingTop'
	| 'paddingX'
	| 'paddingY'
>;

export const BaseButton: FC<IBaseButtonProps> = ({
	align,
	alignSelf,
	backgroundColor,
	borderColor,
	borderRadius,
	borderWidth,
	children,
	color,
	fullHeight = false,
	fullWidth = false,
	grow,
	height,
	justify,
	margin,
	marginBottom,
	marginLeft,
	marginRight,
	marginTop,
	marginX,
	marginY,
	maxHeight,
	maxWidth,
	minHeight,
	minWidth,
	padding,
	paddingBottom,
	paddingLeft,
	paddingRight,
	paddingTop,
	paddingX,
	paddingY,
	shrink,
	width,
	...props
}) => {
	const theme = useEasyFlexTheme();

	const processedBackgroundColor = useColor(backgroundColor, 'transparent');

	const processedBorderColor = useColor(borderColor, undefined);

	const processedBorderRadius = useMemo<string | undefined>(
		() => ifNotUndefined(borderRadius, (borderRadius) => toPx(getBorderRadius(theme, borderRadius))),
		[borderRadius, theme]
	);

	const processedBorderStyle = useMemo<'solid' | undefined>(() => (borderWidth ? 'solid' : undefined), [borderWidth]);

	const processedBorderWidth = useMemo<string | undefined>(
		() => ifNotUndefined(borderWidth, (borderWidth) => toPx(getBorderWidth(theme, borderWidth))),
		[borderWidth, theme]
	);

	const processedColor = useColor(color, undefined);

	const distance = useDistance({
		margin,
		marginBottom,
		marginLeft,
		marginRight,
		marginTop,
		marginX,
		marginY,
		padding,
		paddingBottom,
		paddingLeft,
		paddingRight,
		paddingTop,
		paddingX,
		paddingY,
	});

	const size = useSize({
		fullHeight,
		fullWidth,
		height,
		heightMax: maxHeight,
		heightMin: minHeight,
		width,
		widthMax: maxWidth,
		widthMin: minWidth,
	});

	return (
		<StyledBaseButton
			data-align={align}
			data-align-self={alignSelf}
			data-background-color={processedBackgroundColor}
			data-border-color={processedBorderColor}
			data-border-radius={processedBorderRadius}
			data-border-style={processedBorderStyle}
			data-border-width={processedBorderWidth}
			data-color={processedColor}
			data-grow={grow}
			data-height={size.height}
			data-height-max={size.heightMax}
			data-height-min={size.heightMin}
			data-justify={justify}
			data-margin-bottom={distance.margin.bottom}
			data-margin-left={distance.margin.left}
			data-margin-right={distance.margin.right}
			data-margin-top={distance.margin.top}
			data-padding-bottom={distance.padding.bottom}
			data-padding-left={distance.padding.left}
			data-padding-right={distance.padding.right}
			data-padding-top={distance.padding.top}
			data-shrink={shrink}
			data-width={size.width}
			data-width-max={size.widthMax}
			data-width-min={size.widthMin}
			{...props}
		>
			{children}
		</StyledBaseButton>
	);
};
