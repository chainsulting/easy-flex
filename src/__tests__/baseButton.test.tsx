import { render } from '@testing-library/react';
import 'jest-styled-components';
import React from 'react';
import { BaseButton } from '../baseButton/BaseButton';

describe('<BaseButton />', () => {
	test('should have the basic css settings', async () => {
		const result = render(<BaseButton />);
		expect(result.baseElement).toMatchSnapshot();
	});
});
