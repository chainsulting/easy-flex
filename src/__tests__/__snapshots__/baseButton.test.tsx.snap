// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseButton /> should have the basic css settings 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: none;
  cursor: pointer;
  margin-bottom: 0px;
  margin-left: 0px;
  margin-right: 0px;
  margin-top: 0px;
  padding-bottom: 0px;
  padding-left: 0px;
  padding-right: 0px;
  padding-top: 0px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0:focus:not(:focus-visible) {
  outline: 0;
}

<body>
  <div>
    <button
      class="c0"
      data-direction="row"
      data-margin-bottom="0px"
      data-margin-left="0px"
      data-margin-right="0px"
      data-margin-top="0px"
      data-padding-bottom="0px"
      data-padding-left="0px"
      data-padding-right="0px"
      data-padding-top="0px"
    />
  </div>
</body>
`;
